### DISCUSSION TITLE
Geolocation not working - IPs are not detected.

### DISCUSSION THREAD

**Discussion by monarchmaisuriya** [shlinkio/discussions#171] [Help wanted] [answered]
#### How Shlink is set-up

* Shlink Version:  2.9.2
* PHP Version: 7.4.x
* How do you serve Shlink: Docker image
* Database engine used: PostgreSQL

#### Summary

I was initially using an nginx reverse proxy in front of the shlink instance, but with that the geolocation feature did not work, I would get unknown everytime. So I decided to get rid of the reverse proxy and hit the shlink instance directly, but the geolocation still shows unknown, there is no problem with the geolite configs, the api-key works, the visit commands work. But the geolocation doesn't. I am not sure what I am doing wrong. Can someone point out ?

Just for reference, here is my nginx config -

``` 
server {
    listen 80;
    location / {
        proxy_hide_header 'Access-Control-Allow-Origin';
        add_header 'Access-Control-Allow-Origin' '*' always;
        proxy_set_header Host $host;
        proxy_set_header Referer $http_referer;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Host $remote_addr;
        proxy_set_header X-Forwarded-Referrer $http_referer;
        proxy_set_header X-Forwarded-For $http_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://server:8080;
        proxy_read_timeout 90s;
    }
}
```

And my .env for the docker instance - 

```
# Instance Configs
PORT=8080

# Domain Configs
DEFAULT_DOMAIN=localhost:8080
USE_HTTPS=false

# Postgres Configs
POSTGRES_USER=user
POSTGRES_PASSWORD=pass
POSTGRES_DB=shlink

# Primary Database Configs
DB_DRIVER=postgres
DB_USER=user
DB_PASSWORD=pass
DB_NAME=shlink
# DB_HOST=postgres
DB_HOST=shlink_postgres_1
DB_PORT=5432

# GeoLocation Database Configs
GEOLITE_LICENSE_KEY=123abc

# Docker Specific Configs
ENABLE_PERIODIC_VISIT_LOCATE=true

```

This is the response object for visits -

```
"visitLocation": {
        "countryCode": "",
        "countryName": "",
        "regionName": "",
        "cityName": "",
        "latitude": 0,
        "longitude": 0,
        "timezone": "",
        "isEmpty": true
      },
   ```
   
   I am guessing the `isEmpty` flag being `true` would mean that the IP address could not be captured ?     

PS. This is an amazing project. Huge thanks @acelaya. Great Work!!

**Comment by acelaya** [comment]
Can you check the remote_addr column in the visits table in the database? Does it have any values?

**Comment by monarchmaisuriya** [comment]
> Can you check the remote_addr column in the visits table in the database? Does it have any values?

It has this value - `192.168.240.0` for every visit which is actually the subnet of the docker network I am running the instance on. So I  think I messed up something while setting up the instance with docker. Any clue what might be causing this?

**Reply by acelaya** [reply]
I would say the answer is probably here https://docs.docker.com/network/

Try with different drivers, as by default, bridge is used.

**Reply by monarchmaisuriya** [reply]
I have tried with bridge networks and overlay networks, neither seems to work, but I will keep looking. Anyway thanks, I will update here when I get it working.

**Reply by acelaya** [reply]
Just in case it helps, I also run my own Shlink instances via docker containers, but I have a load balancer/reverse proxy in front of it that runs in the host machine.

By doing so, you can capture the actual IP address there, and then forward it to Shlink.

Your nginx config above should work. Any of these two would do the trick:

```
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-Host $remote_addr;
```

**Reply by monarchmaisuriya** [reply]
Thanks, I am still trying to figure out what's wrong. This nginx config doesn't seem to help.

**Reply by acelaya** [reply]
It won't help if the nginx is also run inside docker, as it will suffer the same networking issue. That's what I meant when I said I run my reverse proxy directly in the host machine. It's not dockerized.

**Reply by monarchmaisuriya** [reply]
Okay, but on my local setup I am directly accessing the shlink instance, no reverse proxy (neither on host nor inside docker container), that should work. If that does not work on its own, I don't think adding a reverse proxy would help. I will still give it a try. Thanks.

**Reply by acelaya** [reply]
It will, if the reverse proxy is outside of docker.

**Reply by monarchmaisuriya** [reply]
Hi, I tried the reverse proxy on my host machine, and this is what I get for `$remote_addr` - `::1`, using the same config mentioned above. I did get an IP when using another device on the same network, but even after running - `shlink visit:locate` the location still shows unknown.

**Reply by AbdelrahmanHafez** [reply]
@ingeniousambivert @acelaya 
Did you have any luck with that? I'm facing the same issue.

* nginx on host machine, I have the following config:
```
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
```
* dockerized shlink with MariaDB

Running:
```sh
docker-compose exec backend shlink visit:locate --retry
```

Outputs:
```txt
...
Processing IP 192.168.192.0 [Could not locate address]
Processing IP 192.168.192.0 [Could not locate address]
Processing IP 192.168.192.0 [Could not locate address]
...
```

I can see that in `config/autoload/client-detection.global.php` we have 
```php
[

    'ip_address_resolution' => [
        'headers_to_inspect' => [
            'CF-Connecting-IP',
            'X-Forwarded-For',
            'X-Forwarded',
            'Forwarded',
            'True-Client-IP',
            'X-Real-IP',
            'X-Cluster-Client-Ip',
            'Client-Ip',
        ]
    ]
]
```
and would assume that shlink would try to find one of these headers before falling back to the docker ip, but that doesn't seem to be the case.

**Reply by AbdelrahmanHafez** [reply, marked as answer]
For those who are still facing the same issue, one way to solve it is by using `ngx_http_realip_module` which creates a new request that takes the actual user's IP and redirects it to Docker.

in `/etc/nginx/nginx.conf` add:
```
        set_real_ip_from 127.0.0.1;
        real_ip_header X-Forwarded-For;
        real_ip_recursive on;
```

And inside your `sites-enabled` file, for example `/etc/nginx/sites-enabled/myshortwebsite.com`

```
server {
  listen 80;
  server_name myshortwebsite.com;

  location / {
    proxy_pass http://localhost:8080;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }
}

server {
  listen 443 ssl;
  server_name myshortwebsite.com;
  ssl_certificate /etc/letsencrypt/live/myshortwebsite.com/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/myshortwebsite.com/privkey.pem;
  location / {
    proxy_pass http://localhost:8080/;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }
}
```

That way, a request comes from the internet to your server with the real IP, nginx takes it and re-creates the request with the actual IP by looking at `X-Forwarded-For`, which is then redirected to your Shlink backend.

**Comment by DaNiElE-1999** [comment]
If your remote_addr table is already populated with visitor IPs, the most likely cause is that you haven’t set the GEOLITE_LICENSE_KEY environment variable in your container. You can find the details in the docs here:
https://shlink.io/documentation/environment-variables/#geolite_license_key

Without a valid key (or if the shared key has hit its download limit), Shlink will be unable to geolocate visits, so all entries will show up as “Unknown.”

**Reply by acelaya** [reply]
> or if the shared key has hit its download limit

That's not correct. If Shlink has a GeoLite DB instance, visits can still be located, but they will be located with the slightly older version of the database.

The download limit only affects downloading newer versions of the database, which Shlink attempts once a month.
